### ISSUE TITLE
[3.9] Warn developers that toolbar.php has been removed

### ISSUE THREAD

**Issue by joeforjoomla** [joomla/joomla-cms#21960]
### Steps to reproduce the issue
Having an extension that use a:

require_once JPATH_ADMINISTRATOR . '/includes/toolbar.php';

### Expected result

JToolbarHelper class is loaded

### Actual result

Fatal error because the file toolbar.php has been removed

### System information (as much as possible)

Joomla 3.9 Alpha

**Comment by zero-24** [comment]
hmm maybe we need to place a file that is still loading this class with an deprecation notice until 4.0 as this could be counted as a B/C? cc @mbabker @wilsonge ?

**Comment by joeforjoomla** [comment]
That would be even better @zero-24

**Comment by mbabker** [comment]
What is the use case where you are including this file?  Before 3.9 it was expected that it was only available in the admin application (and appropriately manually loaded in that case), the class/file was not expected to be used in any other case.

Generally, we don't make file locations a B/C thing, otherwise everything that was namespaced in 3.8 would have to have kept stub files for this exact same reason.

It kind of sounds like your `require_once` line isn't wrapped in a `class_exists('JToolbarHelper')` check, which would be the safe thing to do in the first place and have prevented the error in your case.

**Comment by laoneo** [comment]
It is also recommended to use JLoader to include files which will not crash when the file doesn't exist.

**Comment by joeforjoomla** [comment]
I agree but it's not my code, just found it in another extension. So i'm concerned that there is someone out there using a require_once to include that file and use JToolbarHelper in the frontend for example.

**Comment by softforge** [comment]
I am getting this issue when I update 3.8.12 to 3.9beta3 or beta4 and try to use JCE in the backend
2018-10-07 11:42:43	Warning	82.36.19.205		mod_fcgid: stderr: PHP Warning: require_once(/var/www/vhosts/sfdemo.co.uk/nightly.sfdemo.co.uk/administrator/includes/toolbar.php): failed to open stream: No such file or directory in /var/www/vhosts/sfdemo.co.uk/nightly.sfdemo.co.uk/administrator/components/com_jce/controller.php on line 134				Apache error
2018-10-07 11:42:43	Warning	82.36.19.205		mod_fcgid: stderr: PHP Fatal error: require_once(): Failed opening required '/var/www/vhosts/sfdemo.co.uk/nightly.sfdemo.co.uk/administrator/includes/toolbar.php' (include_path='.:/opt/plesk/php/7.2/share/pear') in /var/www/vhosts/sfdemo.co.uk/nightly.sfdemo.co.uk/administrator/components/com_jce/controller.php on line 134

Sandra is not so may be something to do with my setup<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21960">issues.joomla.org/tracker/joomla-cms/21960</a>.</sub>

**Comment by joeforjoomla** [comment]
Just a few hours after the release of 3.9 and it happened what i said you some months ago guys #22874

**Comment by EArgen** [comment]
I just updated to 3.9.0 and am receiving this error on one of my websites. I can't use the admin front end and simply see this error:

An error has occurred.
0 Class 'JToolbarHelper' not found
 Return to Control Panel

But when I click the button 'Return to Control Panel' nothing happens. What should I do?

**Comment by vz82** [comment]
> I just updated to 3.9.0 and am receiving this error on one of my websites. I can't use the admin front end and simply see this error:
> 
> An error has occurred.
> 0 Class 'JToolbarHelper' not found
> Return to Control Panel
> 
> But when I click the button 'Return to Control Panel' nothing happens. What should I do?

Hi there!

Try this workaround: 
just replace the index.php in your administrator/includes with the most recent version from the backup you've (of course! ;) ) previously made.

**Comment by EArgen** [comment]
Hi there vz82. I'll give that a try! 

Just so you know, the link you provided doesn't seem to work.

**Comment by vz82** [comment]
> Hi there vz82. I'll give that a try!
> 
> Just so you know, the link you provided doesn't seem to work.

Yeah, I've just noticed... :)

Good luck!

**Comment by EArgen** [comment]
Strange, in my backup folder:
(website\html\administrator\includes)
There is no index.php.

**Comment by vz82** [comment]
> Strange, in my backup folder:
> (website\html\administrator\includes)
> There is no index.php.

Try (only) administrator/ then...

**Comment by EArgen** [comment]
That did the trick!
Now of course it's stating that I need to update to Joomla 3.9.0. 

Looking closer, this is what the backup file had:

require_once JPATH_BASE . '/includes/toolbar.php';

Whereas this is that same line on the new version:

require_once JPATH_BASE . '/includes/subtoolbar.php';

**Comment by mbabker** [comment]
https://github.com/joomla/joomla-cms/pull/21372 moved the file to an autoloaded class path and namespaced the class, the same as was done for 300 classes in 3.8.

Check to ensure the `libraries/classmap.php` file matches the 3.9.0 package, that handles ensuring the `JToolbarHelper` class name is registered to the autoloader and it should be pretty close to impossible for that class name to not be able to be resolved unless there is something overloading what we're pushing into the autoloader (since that is set up very early in the process, before any extension even gets loaded).

**Comment by mbabker** [comment]
And to be honest I don't think this "class not found" error has anything to do with the original file having been moved.  Instead, it seems like we're dealing with an autoloader conflict at a first glance, and if that's the case we need to know more about the operating environments where this issue is coming up (i.e. installed extensions).

**Comment by er10077** [comment]
Just replace  require_once JPATH_ADMINISTRATOR . '/includes/toolbar.php';
with
jimport('joomla.html.toolbar');
and it works again

**Comment by mbabker** [comment]
> replace require_once JPATH_ADMINISTRATOR . '/includes/toolbar.php';

If you still have that line in a core file after the 3.9 update then the update did not run correctly.

> with
> jimport('joomla.html.toolbar');

Bad advice.  That too is attempting to load a non-existing file (and if you have that file in a 3.x site you've got other problems).

**Comment by rkbundy** [comment]
VZ82, you are fantastic!  Thank you for the work-around. I don't know how much I'll have to dig into this at night to make sure 3.9.0 is working fully, but at least I can get to the back-end again.

**Comment by er10077** [comment]
So, the correct way for now is "require_once JPATH_ADMINISTRATOR . '/includes/subtoolbar.php';" right?
I don't have file "toolbar.php" in  JPATH_ADMINISTRATOR . '/includes/toolbar.php'; , but jimport('joomla.html.toolbar'); works too. I don't known from where exactly it comes, maybe from /libraries/src/Toolbar/Toolbar.php ?

**Comment by laoneo** [comment]
Turn on debug mode in the joomla config. What shows the stack trace?

**Comment by mbabker** [comment]
> So, the correct way for now is "require_once JPATH_ADMINISTRATOR . '/includes/subtoolbar.php';" right?

No.  That's a completely different file (that was being loaded by the previous `administrator/includes/toolbar.php` file).

> I don't have file "toolbar.php" in JPATH_ADMINISTRATOR . '/includes/toolbar.php'; , but jimport('joomla.html.toolbar'); works too. I don't known from where exactly it comes, maybe from /libraries/src/Toolbar/Toolbar.php ?

`jimport('joomla.html.toolbar');` tries to import a file at `libraries/joomla/html/toolbar.php`.  That is where the JToolbar class existed in 2.5 and was moved to `libraries/cms/toolbar/toolbar.php` at 3.0.  It is still a separate file/class from the JToolbarHelper class being discussed here.

**Comment by er10077** [comment]
ok, thank you. It looks, for now, don't need to include anything to get toolbar working in the frontend of the component. Great work.

Turn on debug mode in the Joomla config. What shows the stack trace?

I don't find anything strange in the stack trace.

**Comment by WebSpain** [comment]
I got hit with this on the 3.9 upgrade and it left the Autostand component inaccessible on the front end.

```
Warning: require_once(/home/findspa/public_html/cars2/administrator/includes/toolbar.php): failed to open stream: No such file or directory in /home/findspa/public_html/cars2/components/com_autostand/autostand.php on line 20

Fatal error: require_once(): Failed opening required '/home/findspa/public_html/cars2/administrator/includes/toolbar.php' (include_path='.:/opt/cpanel/ea-php72/root/usr/share/pear') in /home/findspa/public_html/cars2/components/com_autostand/autostand.php on line 20
```
I put in the toolbar.php from 3.8.13 and it loads again.......but it seems like Autostand will need to be modded going forward.

**Comment by er10077** [comment]
Did you try just remove this "require_once" line? For me, it works.

**Comment by WebSpain** [comment]
Yes.....and it loads fine.

But I wanted to get this one out there so other users of the component will be aware of the issue.

The developers of this application are not generally quick at responding to queries......so there'll be some blitzed Autostands over the next few days.

**Comment by rkbundy** [comment]
More information from here.  Installed the upgrade on one customer site - worked fine, no problems.

Installed on another site and got the issue listed in this thread (toolbar.php & subtoolbar.php). Got rid of subtoolbar - and could access the back-end again. Whew.

Go to "Articles" and get error "Unknown column 'a.note' in 'field list' ".  Turned on debug mode and got a big stack of stuff.  

Went over to the upgraded website that DID work - saw in the upgrade log several SQL statements that were run as part of the upgrade including ALTER stmt to add "Note" field.  So apparently, the upgrade didn't get that far. 

As noted above by @EArgen, my site says that it still needs to upgrade to 3.9.0 (System Information still says I'm at 3.8.13).  Completely confused and befuddled at this point. Even though I've been managing Joomla sites for 5+ years, any of you are much more proficient on the back-end.  

Looking for suggestions - should I re-run the upgrade?  If it thinks I'm at 3.8.13, will running it again (and hopefully finishing correctly) be the right strategy?  Any suggestions on this subject are appreciated.  Thank you very much.

**Comment by vz82** [comment]
> VZ82, you are fantastic! Thank you for the work-around. I don't know how much I'll have to dig into this > at night to make sure 3.9.0 is working fully, but at least I can get to the back-end again.

Tnx! That (getting to be able to access the back-end) was my primary goal anyway, as I was just preparing to add some articles to my page.

As you've mentioned below, my upgrade also seemed to be stuck somewhere,,,

I've ended up using Akeeba to restore to the previous version (now it's showing that 3.9.0 Update is available again; the only issue I can see is with Global Conf. not being formatted properly, I'll try to fix that manually) and I'll just wait until I see any other updates available (then do a backup first, both with Akeeba and via cPanel, as I don't have time for debugging / fixing it).

Good luck anyway!

**Comment by tonypartridge** [comment]
Just found this issue too, checked the file and there was no deprecation notice at all in the file? How was this removed without deprecating it first? 

Themler for one uses it as I've noticed on a clients site.

**Comment by alikon** [comment]
what's wrong on test alpha,beta,RC....

**Comment by mbabker** [comment]
Files in the filesystem do not get deprecations.  Otherwise all ~400 classes that have been namespaced would still have to have stub files in their old locations.  The ONLY exception that has been made is with the old files for the PHPUTF8 library and that is because that library is made up of functions, not classes, and functions do not have a method to be autoloaded so the only way to use that library was by manually including the relevant files (in the Framework's String package we list all the files in the Composer manifest's autoload section so that when the Composer autoloader is imported each of the files is immediately loaded as well).

The `JToolbarHelper` class was only designed to be loaded in the admin backend and was manually included in the bootstrap.  Which means extensions running into this issue:

- Were loading it in the frontend
- Were not using a `class_exists('JToolbarHelper')` check to ensure something else in the process hadn't included it first

**Comment by tonypartridge** [comment]
I agree, but it's Themler using it not me. Where I've used it in the past it was never required. 

I've tested and all works pretty well. Just rather surprised this was removed in a 3.X series. 

It's just the thousands of sites using themler that are going to have issues because if it, I dont like Themler... But I know lots of non techy people use it.

**Comment by alikon** [comment]
me silly ....still dreaming about the ecosystem is in synch with the system

**Comment by Jofrid** [comment]
Hi, this is what I see when I try to log into the editor of my Joomla site: 
An error has occurred.
0 Class 'JToolbarHelper' not found
As described by others above. 

Is there ANY way to fix this?
I have a backup. But how do I get to it?

Thanks.

**Comment by tonypartridge** [comment]
You should be able to just add the missing toolbar.php file to the administrator/includes/ folder as a quick fix.

But enabling debug mode and finding the offending module would be best.

I’ve noticed jomclassifieds breaking in this issue too as they add a custom menu module to the whole admin backend.

**Comment by Jofrid** [comment]
Tony, thanks for answering. What/where is "the missing toolbar.php file"?

**Comment by tonypartridge** [comment]
In your backup? Or any 3.8.x package. In:

administrator/includes/toolbar.php

**Comment by Jofrid** [comment]
My most recent backup is just on my site - which I cannot get to.
I have an older backup downloaded to my PC, but no idea how to use that to restore my website.

Should have gone with wordpress maybe :-)

**Comment by Jofrid** [comment]
No matter where I click or what I do I see only the same error message.

**Comment by tonypartridge** [comment]
You need to access you website files to restore the file as a temporary fix not via the Joomla! admin. But the file restore is a temporary fix as mentioned, enabling Joomla! Debug in the configuration.php file should show the call stack and the offending file trying to include the class.

The reason this has occurred is due to poor coding of a module you are using. Regardless of the CMS a similar situation can happen.

**Comment by Jofrid** [comment]
Thank you again for answering, but as you might already have guessed I do not have enough coding knowledge to understand what you are asking me to do. You tell me to "access my website files to restore the fix". Do you mean upload an older version of toolbar.php from my PC by using an ftp uploader?

**Comment by tonypartridge** [comment]
Yes that would work

**Comment by Jofrid** [comment]
Sadly problem is still there. I have a backup made with Akeeba but no idea how to restore my site with it. Situation is pretty grim...

**Comment by tonypartridge** [comment]
Have you enabled debug mode yet within the configuration.php file? 1 = on 0 = off

**Comment by Jofrid** [comment]
I can't get to any file on the site. Just get the error message. I have the entire site downloaded to my PC, but that is a 1,5year old version.

**Comment by Jofrid** [comment]
I changed the 0 in the configuration.php file on my PC to 1 and uploaded. Now I have a new error message, saying: No configuration file found and no installation code available. Exiting...

**Comment by tonypartridge** [comment]
That sounds like you edited the file name opposed to the contents....

**Comment by Jofrid** [comment]
Thanks for trying to help Tony, very appreciated. 

I ended up restoring from an earlier downloaded backup, by uploading the backup together with Akeeba kickstar and follow the instructions.

Lost a lot of content of course, but at least the site works again.

 

Now – the joomla update 3.9 has presented itself again. What to do about that?

Ignore it and wait for next update?

 

Mvh,

 

Jofrid Bjelland

**Comment by tonypartridge** [comment]
The file won’t be re-added in additional releases, you need to find the module/plugin using the file. Make sure all extensions etc are up to date.

**Comment by Jofrid** [comment]
OK, so I understand the problem is not solved.

How do I identify this file?

 

 

Mvh,

 

Jofrid Bjelland

**Comment by tonypartridge** [comment]
You likely need to seek help from a developer. There is a 3rd party module/plugin you have installed which is causing it.

You would need to clone the site and debug it, like 20minutes work for a developer to find the cause.

**Comment by Jofrid** [comment]
Dzeezes.
Could i disable suspected modules and try (if it fails again I will reinstall again)? Or do I have to uninstall  suspected modules?
Mvh
Jofrid Bjelland

**Comment by laoneo** [comment]
If you create an empty file in administrator/includes/toolbar.php, does the error not go away? Enable the debug plugin, then a stack trace will be displayed which should guide you to the extension. If no stack trace is shown, then open the site for a test with the protostar template. Some templates have a broken error.php file.

**Comment by vz82** [comment]
Hi everyone again,

I'm pretty much sure that it was **T3 Framework** causing the problems (it was updated on November 1 and the upgrade has just succeeded with no errors whatsoever).

Therefore, just make sure to update it BEFORE the Joomla! update (and have a valid backup - of course - just in case :) ).

**Comment by FuriousCamel** [comment]
This is the stack I get debugging. Uploading the 3.8.13 toolbar.php file did not work... Any help on how to solve (back-up not available)?

Call stack
--
# | Function | Location
1 | () | JROOT/administrator/components/com_cpanel/views/cpanel/view.html.php:36
2 | CpanelViewCpanel->display() | JROOT/libraries/src/MVC/Controller/BaseController.php:672
3 | Joomla\CMS\MVC\Controller\BaseController->display() | JROOT/libraries/src/MVC/Controller/BaseController.php:710
4 | Joomla\CMS\MVC\Controller\BaseController->execute() | JROOT/administrator/components/com_cpanel/cpanel.php:15
5 | require_once() | JROOT/libraries/src/Component/ComponentHelper.php:382
6 | Joomla\CMS\Component\ComponentHelper::executeComponent() | JROOT/libraries/src/Component/ComponentHelper.php:357
7 | Joomla\CMS\Component\ComponentHelper::renderComponent() | JROOT/libraries/src/Application/AdministratorApplication.php:101
8 | Joomla\CMS\Application\AdministratorApplication->dispatch() | JROOT/libraries/src/Application/AdministratorApplication.php:159
9 | Joomla\CMS\Application\AdministratorApplication->doExecute() | JROOT/libraries/src/Application/CMSApplication.php:195
10 | Joomla\CMS\Application\CMSApplication->execute() | JROOT/administrator/index.php:51

**Comment by vz82** [comment]
> This is the stack I get debugging. Uploading the 3.8.13 toolbar.php file did not work... Any help on how > to solve (back-up not available)?
> 

Have you tried uploading the index.php file (as mentioned earlier)?

**Comment by sunilongithub** [comment]
Hello, I was able to fix this issue with one of my site by editing following file:
/sitefolder/administrator/includes/helper.php

I added following code to the helper.php:

if(!class_exists('JToolbarHelper')) {
   require_once JPATH_ADMINISTRATOR . '/includes/toolbar.php';
}


above this line in that file:
class JAdministratorHelper

I was able to update it again to latest version. I feel it was caused by poor internet connectivity with the hosting server.

Hope this helps someone here.
